### ISSUE TITLE
Spring Cloud AWS SQS - some messages do not get acknowledged

### ISSUE THREAD

**Issue by anna-kr** [awspring/spring-cloud-aws#1029]
**Type**: Bug

**Component**:
SQS

**Describe the bug**
JDK:17
spring-cloud-aws-dependencies : 3.1.0
org.springframework.boot: 3.2.2

When using `BatchingAcknowledgementProcessor` some messages do not get acknowledged and end up in a dead letter queue even if the processing went without any errors. Looking at the trace logs it seems that there is some race condition between purging `acksBuffer`  and adding new items to it:
`Jan 24, 2024 @ 20:09:37.170 Received message ${messageId} to ack in io.awspring.cloud.sqs.sqsListenerEndpointContainer#2-0.`
`Jan 24, 2024 @ 20:09:37.170 Removing groups [default] from buffer in io.awspring.cloud.sqs.sqsListenerEndpointContainer#2-0`

**Sample**
```
    @Bean
    fun defaultSqsListenerContainerFactory(sqsAsyncClient: SqsAsyncClient): SqsMessageListenerContainerFactory<Any> {
        return SqsMessageListenerContainerFactory.builder<Any>()
            .configure { options -> options
                .acknowledgementMode(AcknowledgementMode.MANUAL)
            }
            .sqsAsyncClient(sqsAsyncClient)
            .build()
    }

  @SqsListener(queueNames = "queue-name")
    fun process(
        sqsMessage: SqsMessage,
        acknowledgement: Acknowledgement
    ) {
        process()
       acknowledgment.acknowledge()
    }
```

**Comment by tomazfernandes** [comment]
Thanks for reporting @anna-kr. 

This is a duplicate of https://github.com/awspring/spring-cloud-aws/issues/925 so I'll close this one.

Just to make sure I understand correctly your use-case - is your queue configured to send messages to DLQ after first attempt?

Thanks.

**Comment by tomazfernandes** [comment]
Duplicate of https://github.com/awspring/spring-cloud-aws/issues/925.

**Comment by anna-kr** [comment]
I am not sure whether this is a duplicate as the issue does not happen during graceful shutdown. I have also observed the problem during graceful shutdown, but was able to mitigate it with delaying batch processor shutdown. Yes, the queue is configured to send messages to DLQ after first attempt.

**Comment by tomazfernandes** [comment]
Oh, I see, my mistake, I'll reopen then.

Can you share more details on your investigation if you have any?

Thanks.

**Comment by anna-kr** [comment]
I haven't done more investigation then what was put into description - basically that messages ending up in DLQ coincide in time with logs about purging `acksBuffer`. It would seem that one thread is adding a message to `acksBuffer` while the other purges the buffer content.

**Comment by tomazfernandes** [comment]
Hi @anna-kr, you were right, and there was indeed a missing part of synchronization in acknowledgement buffer access that made it possible for messages to be removed while the other thread was adding them.

I wasn't able to reproduce the issue with my tests, even testing with a few million messages and 1000 threads, but that's what racing conditions look like anyway.

If you'd like to review the PR, I should keep it open for a few days:
https://github.com/awspring/spring-cloud-aws/pull/1038

Thanks.

**Comment by anna-kr** [comment]
could you tell me when this fix is likely to be released 🙏 ?

**Comment by tomazfernandes** [comment]
Hi @anna-kr - not sure - I still have to go through the issue about waiting on acknowledgements on shutdown and haven't found the time yet.

But hopefully soon!
